src/components/Grid/Grid.tsx

```diff
@@ -0,0 +1,140 @@
+import React, {
+  Children as ReactChildren,
+  createElement,
+  cloneElement,
+  CSSProperties,
+  Fragment,
+  useEffect,
+  useMemo,
+  useState,
+  createRef,
+  PropsWithChildren,
+} from "react";
+import { forwardRef } from "react";
+import { Children } from "~/utils/typing/children";
+import { SPACE_UNIT } from "~/utils/constants";
+import { Box } from "~/index";
+
+export type ContainerProps = PropsWithChildren<{
+  "data-test-id"?: string;
+  rows: number;
+  cols: number;
+  rowsGap?: number;
+  colsGap?: number;
+  className?: string;
+  style?: CSSProperties;
+}> & {
+  children: React.ReactElement | React.ReactElement[];
+};
+
+const Container = forwardRef<HTMLDivElement, ContainerProps>((props, ref) => {
+  const { children } = props;
+  const [invalidError, setInvalidError] = useState(false);
+  const errorMessage =
+    "invalid child type, all children should be <Grid.Item />";
+
+  const cloned = useMemo(
+    () =>
+      React.Children.map(children, (child) =>
+        cloneElement(child, { ref: createRef() })
+      ),
+    [children]
+  );
+
+  useEffect(() => {
+    if (Array.isArray(cloned)) {
+      let count = 0;
+      cloned.forEach((child) => {
+        if (
+          child.ref?.current.dataset.check &&
+          child.ref?.current.dataset.check === "grid-item"
+        )
+          count++;
+      });
+      if (cloned.length !== count) {
+        console.error(errorMessage);
+        setInvalidError(true);
+      } else {
+        setInvalidError(false);
+      }
+    }
+  }, [cloned]);
+
+  const cssStyle: CSSProperties = useMemo(
+    () => ({
+      display: "grid",
+      gridTemplateColumns: `repeat(${props.cols}, 1fr)`,
+      gridTemplateRows: `repeat(${props.rows}, 1fr)`,
+      columnGap: SPACE_UNIT * (props.colsGap || 0),
+      rowGap: SPACE_UNIT * (props.rowsGap || 0),
+      ...props.style,
+    }),
+    [props.rows, props.rowsGap, props.cols, props.colsGap, props.style]
+  );
+
+  return (
+    <div
+      ref={ref}
+      style={cssStyle}
+      data-test-id={props["data-test-id"]}
+      className={props.className}
+    >
+      {invalidError ? (
+        <div>{errorMessage}</div>
```

Comment:
please do not block the UI just print a warning in the console, this seems to much closed as behaviour

Reply:
I really want to catch the implementation errors, why the UI should render incorrectly if we can block it?

Reply:
because there could be cases when you could need this component to be open for expansion. Or someone could use this and don't like the `Container -> Item` relation for whatever reason, it seems just a common behaviour to trigger a warning, maybe add some information in the warning, like the file where the rule is not applied, but it seems too much imperative to block the Ui

Reply:
ok, I just leave the console.warn message